### PR TITLE
RFC: add pytree-compatible dataclass decorator

### PR DESCRIPTION
**Pull request by jakevdp**
I think it's time for JAX to have its own pytree-compatible `dataclass` decorator similar to `flax.struct.dataclass` and `chex.dataclass`. These are mechanisms that have been re-invented by several downstream packages, and JAX can serve users well by providing it directly in JAX itself.

Given the (relatively recent) addition of `register_dataclass`, the implementation is quite concise.

Comments welcome!

Example usage:
```python
import jax

@jax.tree.dataclass
class MyClass:
    arrs: list[jax.Array]                    # By default, attributes will be treated as data fields.
    name: str = jax.tree.field(static=True)  # Optionally mark attributes as static via extended
                                             #  `dataclasses.field` mechanism.

m = MyClass([jax.numpy.arange(4), jax.numpy.ones(2)], "my class")

print(jax.tree.structure(m))
# PyTreeDef(CustomNode(MyClass[('my class',)], [[*, *]]))

print(jax.tree.leaves(m))
# [Array([0, 1, 2, 3], dtype=int32), Array([1., 1.], dtype=float32)]
```
Notice that the metadata specification uses Python's standard `dataclasses.field` mechanism. Rather than the `pytree_node=False` nomenclature used by `flax.struct.dataclass`, I opted for `static=True`, because it makes more clear the connection between pytree metadata and static arguments to JIT.

### CODE REVIEW

**Comment by mattjj** [comment]
Good idea, motivation, and execution!

File: jax/_src/tree.py [hunk 1]
```diff
@@ -14,7 +14,9 @@
 from __future__ import annotations
 
 from collections.abc import Callable, Iterable
+import dataclasses
 from typing import Any, TypeVar, overload
+from typing_extensions import dataclass_transform
```
Comment (yashk2810) [inline jax/_src/tree.py:19]: Isn't this available via typing?

https://peps.python.org/pep-0681/

This would allow to get rid of the typing_extensions dependency or atleast make it conditional on python version if typing.dataclass_transform is only available after certain python versions.

File: jax/_src/tree.py [hunk 2]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
   """
   return tree_util.tree_unflatten(treedef, leaves)
+
+
+def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwds) -> dataclasses.Field:
+  """Used to mark a jax.dataclass field as static.
+
+  See {func}`jax.tree.dataclass` for details.
+  """
+  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
```
Comment (mattjj) [inline jax/_src/tree.py:296]: Should we check that 'static' isn't in `metadata`?

File: jax/_src/tree.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/tree.py:19]: It is starting in Python 3.11, but we still need to support Python 3.10.

Comment (yashk2810) [reply jax/_src/tree.py:19]: So make the `typing_extension` requirement python version dependent and make the import dependent on `sys.version`?

It's better to not include packages by default if we don't need them and it's very simple to do that here. WDYT?

Comment (jakevdp) [reply jax/_src/tree.py:19]: Oh, but `frozen_default` is only included starting in Python 3.12. We could certainly put logic for this in our code, but the point of `typing_extensions` is to take care of this kind of logic for you, and I suspect most users already have it installed (e.g. flax requires it).

Comment (yashk2810) [reply jax/_src/tree.py:19]: Hmm.. okay. Do we need to use that field? Can't we just freeze it by default inside like what you are doing right now? `kwargs.setdefault('frozen', True)`

Comment (jakevdp) [reply jax/_src/tree.py:19]: Yeah, but then the static type checker doesn't know that the result is frozen.

Comment (jakevdp) [reply jax/_src/tree.py:19]: In the end that's a *very* small improvement to the type checker's information, so if you'd prefer to leave it out in order to simplify the deps, that's reasonable I think.

File: jax/_src/tree.py [hunk 3]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
+
+
+@dataclass_transform(field_specifiers=(field, dataclasses.Field), frozen_default=True)
+def dataclass(cls: T, /, **kwargs) -> T:
+  """Create a pytree-compatible dataclass.
```
Comment (superbobry) [inline jax/_src/tree.py:301]: Is the fact that we return a standard library dataclass part of our public API contract? E.g. can users assume that `dataclasses.replace` just works and `dataclasses.is_dataclass` returns True?

File: jax/_src/tree.py [hunk 4]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
+
+def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwds) -> dataclasses.Field:
+  """Used to mark a jax.dataclass field as static.
+
+  See {func}`jax.tree.dataclass` for details.
+  """
+  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
+
+
+@dataclass_transform(field_specifiers=(field, dataclasses.Field), frozen_default=True)
```
Comment (superbobry) [inline jax/_src/tree.py:299]: The fact that `frozen` default to `True` could be surprising for `dataclasses.dataclass` users. Should we call that out in the docstring?

Also, would it be bad not to change the defaults wrt to the standard library `dataclass`?

File: jax/_src/tree.py (hunk 3 above)
Comment (jakevdp) [reply jax/_src/tree.py:301]: It does with this implementation – I think that yes, that should be tested and be part of the contract. What do you think?

File: jax/_src/tree.py [hunk 5]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
+    >>> x = OpDescription("add", jnp.arange(4), 10)
+    >>> f(x)
+    Array([10, 11, 12, 13], dtype=int32)
+  """
+  if '_jax_dataclass' in cls.__dict__:
+    return cls
+  kwargs.setdefault('frozen', True)
+  dcls = dataclasses.dataclass(**kwargs)(cls)  # type: ignore
+  dcls._jax_dataclass = True
+  data_fields = [f.name for f in dataclasses.fields(dcls) if not f.metadata.get('static', False)]
```
Comment (levskaya) [inline jax/_src/tree.py:327]: `dcls.replace = dataclasses.replace`  please please please

File: jax/_src/tree.py (hunk 4 above)
Comment (levskaya) [reply jax/_src/tree.py:299]: That's why you add the `replace` method :p  But srsly don't we want to encourage the frozen default for sanity w. all of JAX tracing?

Comment (cgarciae) [reply jax/_src/tree.py:299]: On Flax struct we recently added kwargs and users immediately figured out they could set `frozen=False` and they seem to enjoy it.

File: jax/_src/tree.py (hunk 2 above)
Comment (francois-rozet) [reply jax/_src/tree.py:296]: The following would crash if `'static'` is in `metadata`.

```suggestion
  return dataclasses.field(metadata=dict(**(metadata or {}), static=static), **kwds)
```

File: jax/_src/tree.py [hunk 6]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
+  meta_fields = [f.name for f in dataclasses.fields(dcls) if f.metadata.get('static', False)]
+  tree_util.register_dataclass(dcls, data_fields, meta_fields)
+  return cls
```
Comment (francois-rozet) [inline jax/_src/tree.py:330]:
```suggestion
  return dcls
```

File: jax/_src/tree.py [hunk 7]
```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
...
+def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwds) -> dataclasses.Field:
+  """Used to mark a jax.dataclass field as static.
+
+  See {func}`jax.tree.dataclass` for details.
+  """
+  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
+
+
+@dataclass_transform(field_specifiers=(field, dataclasses.Field), frozen_default=True)
+def dataclass(cls: T, /, **kwargs) -> T:
```
Comment (francois-rozet) [inline jax/_src/tree.py:300]: With this interface, it is not possible to use `@jax.tree.dataclass(...)` as a decorator (while `@dataclasses.dataclass(...)` is valid.

```suggestion
def dataclass(cls: T = None, /, **kwargs) -> T:

    if cls is None:
        return partial(dataclass, **kwargs)
```

**Comment by lebrice** [comment]
Great idea!
Maybe later on, [Annotated types](https://docs.python.org/3/library/typing.html#typing.Annotated) could be amazing with this. For example, there could be an annotated type for `Static`:

```python
import jax

@jax.tree.dataclass
class MyClass:
    arrs: list[jax.Array]
    name: Static[str]
```

File: jax/_src/tree.py [hunk 8]
```diff
@@ -293,7 +293,7 @@ def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwd
 
   See {func}`jax.tree.dataclass` for details.
   """
-  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
+  return dataclasses.field(metadata={**(metadata or {}), 'static': static}, **kwds)
```
Comment (francois-rozet) [inline jax/_src/tree.py:296]: I don't know if this is on purpose, but this does not prevent `metadata` from having a `static` key.

```python
In [1]: {**{"static": True}, "static": False}
Out[1]: {'static': False}

In [2]: dict(**{"static": True}, static=False)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 1
----> 1 dict(**{"static": True}, static=False)

TypeError: dict() got multiple values for keyword argument 'static'
```

File: jax/_src/tree.py (hunk 5 above)
Comment (superbobry) [reply jax/_src/tree.py:342]: Just to re-iterate my earlier argument, there is currently no way of telling Python type checkers that the result of `dataclass_transform` has an extra method. We need intersection types for that.

File: jax/_src/tree.py (hunk 7 above)
Comment (superbobry) [reply jax/_src/tree.py:300]: We will need a separate overload for the `cls: None` case returning a `Callable[..., T]`.

**Comment by superbobry** [comment]
> Great idea! Maybe later on, [Annotated types](https://docs.python.org/3/library/typing.html#typing.Annotated) could be amazing with this. For example, there could be an annotated type for `Static`:
> 
> ```python
> import jax
> 
> @jax.tree.dataclass
> class MyClass:
>     arrs: list[jax.Array]
>     name: Static[str]
> ```

Note that you don't need `jax.tree.dataclass` if you use `Annotated`. Instead, `jax.tree_util.register_dataclass` could aggregate fields tagged with `Static`, if the given type is a (plain stdlib) dataclass.

File: jax/_src/tree.py (hunk 5 above)
Comment (levskaya) [reply jax/_src/tree.py:342]: One way we make type checkers happy in the flax version is by using a base type, roughly:

```python
TNode = TypeVar('TNode', bound='PyTreeNode')
@dataclass_transform(field_specifiers=(field,))  # type: ignore[literal-required]
class PyTreeNode:
  def __init_subclass__(cls, **kwargs):
    dataclass(cls, **kwargs)  # pytype: disable=wrong-arg-types
  def __init__(self, *args, **kwargs):
    # stub for pytype
    raise NotImplementedError
  def replace(self: TNode, **overrides) -> TNode:
    # stub for pytype
    raise NotImplementedError
```
one uses it like
```
class Thing(struct.PyTreeNode):
  blah: Any
  ...
```

I'd just like some convenience decorator or base class that does the common stuff we use _constantly_ in current implementations.  If we want the blessed interface to be stdlib compliant, fine - but I'd love something without a ton of syntactic boilerplate.

File: jax/_src/tree.py (hunk 7 above)
Comment (jakevdp) [reply jax/_src/tree.py:300]: Yeah I deliberately left this out in the initial draft to keep things simple.

File: jax/_src/tree.py (hunk 8 above)
Comment (jakevdp) [reply jax/_src/tree.py:296]: We'll figure out these details before merging – we're a long way from that.

File: jax/_src/tree.py (hunk 5 above)
Comment (superbobry) [reply jax/_src/tree.py:342]: Yeah, actually, I agree that a base class approach would work. I think we prototyped this for Flax at some point?

**Comment by NeilGirdhar** [comment]
What's the purpose of this? 

```dcls.replace = dataclasses.replace```

I think this is poor design.  Using a replace `method` is only slightly more convenient that importing `replace` from either `dataclasses` or `copy`, but it has drawbacks:
* it's not type-checked (whereas the other functions are),
* it shadows any member variable named `replace` (weird behavior), and
* it makes it a bit harder to convert code that works with `jax.tree.dataclass` to an ordinary dataclass.

**Comment by NeilGirdhar** [comment]
Also, will this be used in Jax's codebase?  From what I remember of my contributions to Jax, it was really hard to follow the many times tuples were packed and unpacked to pass things around.  Using dataclasses more often would be a huge improvement for future contributors :smile:

**Comment by jakevdp** [comment]
> But, is this absolutely necessary?
> 
> `dcls.replace = dataclasses.replace`

I didn't add it initially, but it was requested by users of `flax.struct.dataclass`.

It's still not clear whether we'll add this; currently it's possible to use `@jax.tree_util.register_dataclass` around a built-in `@dataclasses.dataclass` and get basically all the functionality.

**Comment by NeilGirdhar** [comment]
> It's still not clear whether we'll add this; currently it's possible to use `@jax.tree_util.register_dataclass` around a built-in `@dataclasses.dataclass` and get basically all the functionality.

True.  The big benefit is the slightly more compact `field` function when marking something as static, and the default to frozen.  If you do add it, I laboriously copied the type hinting from Python's codebase into my version of dataclasses [here](https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/dataclasses/dataclass.py) and [here](https://github.com/NeilGirdhar/tjax/blob/main/tjax/_src/dataclasses/helpers.py), and you're welcome to lift whatever you like into your code.

Also, if I were doing it again, I would probably switch to a `Dataclass` class like Equinox did mainly because it's easier to keep derived classes consistent (they should also be dataclass-decorated and if they are, they must have the same frozen setting).

**Comment by jakevdp** [comment]
> The big benefit is the slightly more compact `field` function when marking something as static,

That's unrelated to whether we provide a JAX-specific `dataclass` decorator. We specifically decided not to add additional syntax for marking fields in https://github.com/jax-ml/jax/pull/24700

> default to frozen

There was some discussion around whether or not this is a surprising default; it's not clear whether defaulting to frozen is the best choice here. I lean toward yes, but the conservative choice is to stick with the built-in `dataclass` definition and so we've done that for now.

> if I were doing it again, I would probably switch to a `Dataclass` class [like Equinox did](https://github.com/patrick-kidger/equinox/blob/main/equinox/_module.py)

We've discussed this in the past, and in general the JAX team leans toward avoiding inheritance-based design patterns, as they don't fit as well with the functional paradigm that JAX uses.
